modules/transaction-payment/src/tests.rs

```diff
@@ -1041,24 +1043,41 @@ fn swap_from_pool_and_dex_update_rate() {
 		// swap 80 DOT out 3074 ACA
 		let trading_path = Pallet::<Runtime>::get_trading_path_by_currency(&ALICE, DOT).unwrap();
 		let supply_amount = Currencies::free_balance(DOT, &dot_fee_account) - dot_ed;
-		let (_, swap_native) =
+		// here just get swap out amount, the swap not happened
+		let (supply_in_amount, swap_out_native) =
 			module_dex::Pallet::<Runtime>::get_swap_amount(&trading_path, SwapLimit::ExactSupply(supply_amount, 0))
 				.unwrap();
-		assert_eq!(3074, swap_native);
-		// calculate the new balance of ACA = 3074+(init-800)=3074+9200=12270
-		let current_native_balance =
-			(swap_native + Currencies::free_balance(ACA, &dot_fee_account)).saturated_into::<u128>();
-		let base_native: u128 = current_native_balance / 10;
-		assert_eq!(1227, base_native);
-		// compare to the old balance:10000 and rate:1/10, the new rate is (12270/10000)*(1/10)=0.1227
-		let rate = Ratio::saturating_from_rational(base_native, pool_size);
-		assert_eq!(Ratio::saturating_from_rational(1227, 10000), rate);
+		assert_eq!(3074, swap_out_native);
+		assert_eq!(supply_in_amount, supply_amount);
+		let new_pool_size =
+			(swap_out_native + Currencies::free_balance(ACA, &dot_fee_account)).saturated_into::<u128>();
+
+		// the swap also has it's own exchange rate by input_amount divide output_amount
+		let swap_exchange_rate = Ratio::saturating_from_rational(supply_in_amount, swap_out_native);
+		assert_eq!(swap_exchange_rate, Ratio::saturating_from_rational(80, 3074));
+
+		let old_threshold_rate = Ratio::saturating_from_rational(swap_balance_threshold, pool_size);
+		let new_threshold_rate = Ratio::one().saturating_sub(old_threshold_rate);
+
+		let new_exchange_rate = old_exchange_rate
+			.saturating_mul(old_threshold_rate)
+			.saturating_add(swap_exchange_rate.saturating_mul(new_threshold_rate));
```

Comment:
Number literals should be used to verify the math in uni tests. Or you're just repeating the implementation and it will always pass.

Reply:
Also need to add more `swap_amount` cases to cover different scenarios.

Reply:
I use `calculate_exchange_rate` method in pallet, and use it in testcase, this will be better now.

Reply:
No, this is still just repeating the impl, not proper unit tests.

USE NUMBER LITERALS.

Reply:
Let me add more details:

A proper unit tests to verify math should be like

```rust
// impl
fn add_one(num: u32) -> u32 {
  num + 1
}

// tests
#[test]
fn add_one_works() {
  // literal 2 is used
  assert_eq!(add_one(1), 2);
}
```

What you're doing in unit tests currently is similar to:

```rust
// impl
fn add_one(num: u32) -> u32 {
  do_add_one(num)
}

fn do_add_one(num: u32) -> u32 {
  num + 1
}

// tests
#[test]
fn add_one_works() {
  // literal 2 is used
  assert_eq!(add_one(1), do_add_one(1));
}
```

Reply:
👍 done NUMBER LITERALS check in the event.